src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
Since `Entity` is a type alias of a union, it will be narrowed after its assignment. This creates a circular dependency during the type analysis. Because of this, `node` would be typed as `any` if the type annotation was not given.

@ahejlsberg Would it be possible to give the inference of the type of the initializer precedence over the narrowing after assignments?